File: algorithm/study/Core-07-Binary-Search-Hash/2110/python/src/solution.py
import sys
input = sys.stdin.readline

def main():
    N, C = map(int, input().split())
    houses = sorted(int(input()) for _ in range(N))

    def feasible(d):
        """Can we place C routers with min distance >= d?"""
        count = 1
        last = houses[0]
        for i in range(1, N):
            if houses[i] - last >= d:
                count += 1
                last = houses[i]
                if count >= C:
                    return True
        return False

    lo, hi, ans = 1, houses[-1] - houses[0], 0
    while lo <= hi:
        mid = (lo + hi) // 2
        if feasible(mid):
            ans = mid
            lo = mid + 1
        else:
            hi = mid - 1

    print(ans)

if __name__ == "__main__":
    main()

File: algorithm/study/Core-07-Binary-Search-Hash/2110/cpp/src/solution.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int main(){
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int N, C; cin >> N >> C;
    vector<long long> h(N);
    for(int i = 0; i < N; i++) cin >> h[i];
    sort(h.begin(), h.end());

    auto feasible = [&](long long d) -> bool {
        int cnt = 1;
        long long last = h[0];
        for(int i = 1; i < N; i++){
            if(h[i] - last >= d){
                cnt++;
                last = h[i];
                if(cnt >= C) return true;
            }
        }
        return false;
    };

    long long lo = 1, hi = h[N-1] - h[0], ans = 0;
    while(lo <= hi){
        long long mid = (lo + hi) / 2;
        if(feasible(mid)){
            ans = mid;
            lo = mid + 1;
        } else {
            hi = mid - 1;
        }
    }
    cout << ans << "\n";
    return 0;
}
